File: mobile/lib/minio_storage.dart
import 'dart:convert';
import 'app_state.dart' show supabase;

/// Client for the `minio-storage` edge function — uploads/deletes go through
/// it so the MinIO secret key never lives on the device. Same base64-in-JSON
/// shape ai_service.dart already uses for images, so no extra `http`
/// dependency is needed; downloads stay plain public URLs (see the function).
class MinioStorage {
  static Future<String> upload({
    required String bucket,
    required String path,
    required List<int> bytes,
    required String contentType,
  }) async {
    final res = await supabase.functions.invoke(
      'minio-storage',
      body: {
        'action': 'upload',
        'bucket': bucket,
        'path': path,
        'contentType': contentType,
        'base64': base64Encode(bytes),
      },
    );
    final data = res.data;
    if (data is Map && data['error'] != null) throw Exception(data['error']);
    if (data is Map && data['publicUrl'] is String) return data['publicUrl'] as String;
    throw Exception('Unexpected upload response');
  }

  static Future<void> remove(String bucket, List<String> paths) async {
    final res = await supabase.functions.invoke(
      'minio-storage',
      body: {'action': 'delete', 'bucket': bucket, 'paths': paths},
    );
    final data = res.data;
    if (data is Map && data['error'] != null) throw Exception(data['error']);
  }
}

File: src/lib/minioStorage.js
import { supabase } from './supabase';

// All uploads/deletes proxy through the minio-storage edge function so the
// MinIO secret key never ships to the browser — see that function for the
// bucket allowlist and admin-only buckets. Downloads stay direct public URLs.
const FUNCTION_URL = `${import.meta.env.VITE_SUPABASE_URL}/functions/v1/minio-storage`;

async function authHeaders(contentType) {
  const { data } = await supabase.auth.getSession();
  const token = data.session?.access_token;
  if (!token) throw new Error('Not signed in');
  return {
    Authorization: `Bearer ${token}`,
    apikey: import.meta.env.VITE_SUPABASE_ANON_KEY,
    'Content-Type': contentType,
  };
}

// Upload one file, returns the public URL (same shape as
// supabase.storage.from(bucket).getPublicUrl(path).data.publicUrl).
export async function uploadToMinio(bucket, path, file, contentTypeOverride) {
  const contentType = contentTypeOverride || file.type || 'application/octet-stream';
  const headers = await authHeaders(contentType);
  const qs = new URLSearchParams({ action: 'upload', bucket, path, contentType });
  const res = await fetch(`${FUNCTION_URL}?${qs}`, { method: 'POST', headers, body: file });
  const body = await res.json();
  if (!res.ok) throw new Error(body.error || 'Upload failed');
  return body.publicUrl;
}

// Delete one or more objects (same shape as supabase.storage.from(bucket).remove(paths)).
export async function removeFromMinio(bucket, paths) {
  const headers = await authHeaders('application/json');
  const res = await fetch(`${FUNCTION_URL}?action=delete`, {
    method: 'POST',
    headers,
    body: JSON.stringify({ bucket, paths }),
  });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(body.error || 'Delete failed');
  }
}
